### DISCUSSION TITLE
Code completion in FQNs

### DISCUSSION THREAD

**Discussion by goto40** [eclipse-langium/langium#1137] [Q&A] [answered]
I have many references implemented as ['qualified name scoping'](https://langium.org/guides/scoping/qualified-name/). However, I have a usability issue for the vscode plugin, which can easily be demonstrated in a simple example [on the playground](https://langium.org/playground/):

Just change the grammar to use FQNs instead if IDs (introducing dots as part of the name):
```
grammar HelloWorld

entry Model:
    (persons+=Person | greetings+=Greeting)*;

Person:
    'person' name=FQN;

Greeting:
    'Hello' person=[Person:FQN] '!';

FQN returns string: ID ('.' ID)*;

hidden terminal WS: /\s+/;
terminal ID: /[_a-zA-Z][\w_]*/;

hidden terminal ML_COMMENT: /\/\*[\s\S]*?\*\//;
hidden terminal SL_COMMENT: /\/\/[^\n\r]*/;
```

Then my problem gets visible. While typing the start of a name, completion works as expected:
![image](https://github.com/langium/langium/assets/16887130/16398057-d955-4013-9506-2602ed7d3b2c)

Also when typing parts of the first name (before the dot) and parts of the second part of the name (after the dot), completion works as expected:
![image](https://github.com/langium/langium/assets/16887130/c46fbd90-2f3b-4b6f-8843-c273ce989c54)

But when using parts of the first name and a dot, completion breaks at some point:

(1) OK: "Jo.S" --> "John.Simpson", but already "Simpson" is part of the list (w/o first name)
![image](https://github.com/langium/langium/assets/16887130/c84a8baf-754e-4da7-86ee-72f2cc04ebb9)

(2) not OK: "Jo.Si" --> "Simpson"
![image](https://github.com/langium/langium/assets/16887130/12c83771-ad01-4c5e-b133-2faf76660515)

For large models this yields very confusion completion effects :-(

Is there already an issue for that? I did not find one...

**Comment by msujew** [comment, marked as answer]
Hey @goto40,

completion is a pretty complex thing. The current algorithm jumps to the start of the token at the cursor position to identify the position in the grammar. It needs this position to find out which grammar elements can follow the current cursor position. It also has some additional logic in there to deal with assignments and cross reference related elements.

The issue with FQN appears when the start of the token isn't actually the start of the cross reference. The completion would actually need to jump a few more characters back to identify that the current element we're looking at isn't some `ID` inside of an `FQN` rule, but the `person=[Person:FQN]` assignment of the `Greeting` rule.

I've created an PR to fix this in the framework https://github.com/langium/langium/pull/1138.

**Reply by goto40** [reply]
Hi @msujew ,

Thank you for your quick response.

I confirm that tuning the completion provider helps in this issue (with langium 1.2.1):
```
    protected override backtrackToTokenStart(text: string, offset: number): number {
        if(offset>1 && text[offset-1].match(/[\w\d.]/)) offset--;
        while(offset>1 && text[offset-1].match(/[\w\d.]/)) offset--;
        return super.backtrackToTokenStart(text, offset);
    }
```
However, this looks hacky and probably will not work in all cases... I prefer waiting for your fix...

I assume your PR will be part of the 2.0.0 release?
